### PR TITLE
Improve ACL 2.0 authentication and authorization handling

### PR DESCRIPTION
**Pull request by dingshuangxi888**
### Which Issue(s) This PR Fixes

N/A

### Brief Description

Improve ACL 2.0 authentication and authorization handling across remoting, admin, transaction, and gRPC paths.

### How Did You Test This Change?

Ran focused JDK 11 Maven tests across auth, client, broker, tools, proxy, and container modules (222 tests; 221 passed, 1 skipped).

### CODE REVIEW

**Review by RockteMQ-AI — COMMENTED** [review]
/tmp/review_10997/summary.txt

File: auth/src/main/java/org/apache/rocketmq/auth/config/AuthConfig.java [hunk 1]
```diff
@@ -136,11 +183,12 @@ public void setAuthenticationStrategy(String authenticationStrategy) {
     }
 
     public String getAuthenticationWhitelist() {
-        return authenticationWhitelist;
+        WhitelistSnapshot snapshot = authenticationWhitelist;
```
Comment (RockteMQ-AI) [inline auth/src/main/java/org/apache/rocketmq/auth/config/AuthConfig.java:186]: Nice use of `WhitelistSnapshot` as an immutable value object for thread-safe whitelist updates. The `volatile` reference swap ensures visibility across threads without locking.

File: broker/src/main/java/org/apache/rocketmq/broker/processor/AdminBrokerProcessor.java [hunk 2]
```diff
@@ -3031,6 +3063,12 @@ private RemotingCommand resumeCheckHalfMessage(ChannelHandlerContext ctx,
             selectMappedBufferResult = this.brokerController.getMessageStore()
                 .selectOneMessageByOffset(messageId.getOffset());
             MessageExt msg = MessageDecoder.decode(selectMappedBufferResult.getByteBuffer(), true, false);
+            if (!Objects.equals(requestHeader.getTopic(),
+                msg.getUserProperty(MessageConst.PROPERTY_REAL_TOPIC))) {
+                response.setCode(ResponseCode.NO_PERMISSION);
+                response.setRemark("The topic does not match the transaction message");
```
Comment (RockteMQ-AI) [inline broker/src/main/java/org/apache/rocketmq/broker/processor/AdminBrokerProcessor.java:3069]: The topic mismatch check in `resumeCheckHalfMessage` returns `NO_PERMISSION` — consider whether `SYSTEM_ERROR` or a more specific error code would be more appropriate, since this is a topic validation failure rather than a permission denial.

File: broker/src/main/java/org/apache/rocketmq/broker/processor/AdminBrokerProcessor.java [hunk 3]
```diff
@@ -3524,6 +3565,27 @@ private boolean isNotSuperUserLogin(RemotingCommand request) {
...
+        if (this.brokerController.getAuthConfig() == null
+            || !this.brokerController.getAuthConfig()
+            .isAuthenticationRequired(String.valueOf(request.getCode()))) {
+            return false;
+        }
+        String accessKey = getAccessKey(request);
+        if (StringUtils.isEmpty(accessKey)) {
+            return false;
+        }
+        if (StringUtils.equals(accessKey, username)) {
```
Comment (RockteMQ-AI) [inline broker/src/main/java/org/apache/rocketmq/broker/processor/AdminBrokerProcessor.java:3578]: The `canReadUserPassword` logic is well-designed: it requires authentication to be enabled, checks self-access first, then falls back to super-user check. Good defense-in-depth.

File: proxy/src/main/java/org/apache/rocketmq/proxy/grpc/interceptor/HeaderInterceptor.java [hunk 4]
```diff
@@ -58,12 +61,12 @@ public <R, W> ServerCall.Listener<R> interceptCall(
             Metadata.Key<String> headerKey
                     = Metadata.Key.of(key.toString(), Metadata.ASCII_STRING_MARSHALLER);
```
Comment (RockteMQ-AI) [inline proxy/src/main/java/org/apache/rocketmq/proxy/grpc/interceptor/HeaderInterceptor.java:62]: Switching from `putHeaderIfNotExist` to `putHeader` (overwrite) is the correct fix — ensures the proxy authoritative source always wins over client-supplied values.

**Review by RockteMQ-AI — APPROVED** [review]
### Summary
LGTM — comprehensive ACL 2.0 security hardening. The follow-up commit (f1fc31b3) addresses the previous feedback with additional test coverage and edge case handling.

**Key improvements:**
- Proper topic requirement for VIEW_MESSAGE_BY_ID authorization (closes security gap)
- Thread-safe whitelist configuration with immutable snapshots
- gRPC subject trust enforcement in proxy
- Password response restriction (defense-in-depth)
- Comprehensive test coverage for authorization edge cases

The changes are well-structured and the test coverage is thorough.

---
*Automated review by github-manager-bot*

**Review by RongtongJin — APPROVED** [review]

**Review by RockteMQ-AI — APPROVED** [review]
### Summary
Comprehensive ACL 2.0 hardening across authentication, authorization, and compatibility layers. 36 files changed with extensive test coverage.

**Key improvements:**

1. **Thread-safe whitelist management** — `AuthConfig` now owns whitelist parsing via an immutable `WhitelistSnapshot` behind a `volatile` reference. This eliminates the per-strategy mutable `HashSet` that was parsed once at init, enabling safe runtime updates.

2. **Stateful strategy fix** — Whitelisted RPCs now always call `doEvaluate()` instead of returning early. The old code could cache a success on a whitelisted call and then incorrectly bypass authentication for subsequent non-whitelisted RPCs on the same channel. The new test `authenticationWhitelistBypassIsNotCachedAsSuccess` precisely validates this fix.

3. **VIEW_MESSAGE_BY_ID authorization enforced** — Removed from the historical-compatibility bypass in `AuthorizationCompatibility` and now requires a valid topic resource. Previously, blank/missing topics silently skipped authorization — a security gap.

4. **RESUME_CHECK_HALF_MESSAGE authorization added** — New authorization path with proper topic validation, closing a gap where this admin operation had no authz check.

5. **Extensive test coverage** — New test classes (`StatefulAuthenticationStrategyTest`, `AuthConfigTest`) and expanded assertions in existing tests cover the new authorization paths and edge cases.

**Minor observations:**
- The `WhitelistSnapshot` pattern is clean and well-encapsulated
- The `isAuthenticationRequired()` method centralizes the check nicely
- Backward compatibility is preserved for genuinely historical request shapes (END_TRANSACTION without topic)

LGTM — solid security hardening with proper test coverage.

---
*Automated review by github-manager-bot*
